### PR TITLE
Refactor ConstantInterval

### PR DESCRIPTION
**Pull request by abadams**
This is the bulk of #8155, pulled out for separate review.

This PR pulls the ConstantInterval class out of Interval.h, and the arithmetic helpers for it out of Monotonic.cpp. It then refactors Monotonic.cpp to use it, including using a new bounds inference routine to use instead of find_constant_bounds.   It adds fuzz tests for both the new bounds inference routine and the interval arithmetic itself.

The fuzz test shows a lot of errors due to the lossless_cast casting bugs revealed in #8155, so those are simply reported for now, though I did fix the most egregious bug in lossless_cast to reduce the noise. The fuzz test also found another error to fix in FindIntrinsics.cpp

After this, the next step is to refactor the simplifier to use this ConstantInterval instead of rolling its own thing, and then actually fixing the lossless_cast bug, which has trickle-on effects in instruction selection.

### CODE REVIEW

**Comment by abadams** [comment]
@rootjalex I think you would be ideal to review this (Not urgently - once you're out from under your current crunch)

**Comment by rootjalex** [comment]
Will take a look next week!

File: src/ConstantBounds.cpp [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+                return abs(recurse(op->args[0]));
+            } else if (op->is_intrinsic(Call::absd)) {
+                return abs(recurse(op->args[0]) -
+                           recurse(op->args[1]));
+            } else if (op->is_intrinsic(Call::count_leading_zeros) ||
+                       op->is_intrinsic(Call::count_trailing_zeros)) {
+                // Conservatively just say it's the potential number of zeros in the type.
+                return ConstantInterval(0, op->args[0].type().bits());
+            } else if (op->is_intrinsic(Call::halving_add)) {
+                return (recurse(op->args[0]) +
```
Comment (rootjalex) [inline src/ConstantBounds.cpp:91]: I'm not convinced that this computation, or rounding_halving_add below shouldn't have the `cast(op->type, ...)` wrapper. Even if the computation can't overflow, the bounds computation can, just as in halving_sub.

Comment (abadams) [reply src/ConstantBounds.cpp:91]: Yes, it would have to be with a 64-bit type, and you'd get an interval with an infinity bound instead of the bounds of the (64-bit) type. It's not going to cause a miscompilation, but for a uint64 calculation you might for example not exploit the fact that the result is non-negative.

File: src/ConstantInterval.cpp [hunk 2]
```diff
@@ -0,0 +1,643 @@
...
+    if (max_defined) {
+        max = std::max(max, x);
+    }
+    if (min_defined) {
+        min = std::min(min, x);
+    }
+}
+
+bool ConstantInterval::contains(int64_t x) const {
+    return !((min_defined && x < min) ||
```
Comment (rootjalex) [inline src/ConstantInterval.cpp:85]: imo it's much easier to read:
```
(!min_defined || x >= min) && (!max_defined || x <= max)
```

File: src/ConstantInterval.cpp [hunk 3]
```diff
@@ -0,0 +1,643 @@
...
+
+bool ConstantInterval::contains(int32_t x) const {
+    return contains((int64_t)x);
+}
+
+bool ConstantInterval::contains(uint64_t x) const {
+    if (x <= (uint64_t)std::numeric_limits<int64_t>::max()) {
+        return contains((int64_t)x);
+    } else {
+        return !max_defined;
```
Comment (rootjalex) [inline src/ConstantInterval.cpp:97]: The correctness of this line is a bit subtle (because any value of `min` is implicitly less than `x` after the check), and could benefit from a comment explaining it.

File: src/ConstantInterval.cpp [hunk 4]
```diff
@@ -0,0 +1,643 @@
...
+        } else {
+            result.max = a.max;
+        }
+        result.max_defined = true;
+    } else {
+        result.max_defined = b.max_defined;
+        result.max = b.max;
+    }
+    internal_assert(!result.is_bounded() || result.min <= result.max)
+        << "Empty ConstantInterval constructed in make_intersection";
```
Comment (rootjalex) [inline src/ConstantInterval.cpp:133]: Is this error message right? The first branch of the `||` is true if the result is unbounded, i.e. empty?

File: src/ConstantInterval.cpp [hunk 5]
```diff
@@ -0,0 +1,643 @@
...
+                         b.min_defined &&
+                         sub_with_overflow(64, a.max, b.min, &result.max);
+    return result;
+}
+
+ConstantInterval operator/(const ConstantInterval &a, const ConstantInterval &b) {
+    ConstantInterval result;
+
+    result.min = INT64_MAX;
+    result.max = INT64_MIN;
```
Comment (rootjalex) [inline src/ConstantInterval.cpp:267]: Shouldn't these be a.min and a.max? Otherwise I don't think this operator ever returns a bounded interval

Comment (rootjalex) [reply src/ConstantInterval.cpp:267]: Oh, nevermind, I misread. Ignore.

File: src/ConstantInterval.cpp [hunk 6]
```diff
@@ -0,0 +1,643 @@
...
+    }
+
+    // Normalize the values if it's undefined
+    if (!result.min_defined) {
+        result.min = 0;
+    }
+    if (!result.max_defined) {
+        result.max = 0;
+    }
+
```
Comment (rootjalex) [inline src/ConstantInterval.cpp:359]: This function is complex enough that I think having a safety check like in other places would be good.

Comment (rootjalex) [reply src/ConstantInterval.cpp:359]: i.e.
```
internal_assert(!result.is_bounded() || (result.min <= result.max)
```

File: src/ConstantInterval.cpp [hunk 7]
```diff
@@ -0,0 +1,643 @@
...
+    if (!result.min_defined) {
+        result.min = 0;
+    }
+
+    if (!result.max_defined) {
+        result.max = 0;
+    }
+
+    return result;
```
Comment (rootjalex) [inline src/ConstantInterval.cpp:432]: Same safety check here I think

File: src/ConstantInterval.cpp (hunk 4 above)
Comment (abadams) [reply src/ConstantInterval.cpp:133]: unbounded means goes to infinity on at least one side, not empty. I'm not sure how we'd represent empty, but I think it would be min_defined && max_defined && min > max.

File: src/ConstantInterval.cpp [hunk 8]
```diff
@@ -0,0 +1,660 @@
...
+
+    // Mod by produces a result between 0
+    // and max(0, abs(modulus) - 1). However, if b is unbounded in
+    // either direction, abs(modulus) could be arbitrarily
+    // large.
+    if (b.is_bounded()) {
+        result.max_defined = true;
+        result.max = 0;                                 // When b == 0
+        result.max = std::max(result.max, b.max - 1);   // When b > 0
+        result.max = std::max(result.max, -1 - b.min);  // When b < 0
```
Comment (rootjalex) [inline src/ConstantInterval.cpp:469]: What happens if these computations overflow? i.e. if b is a single point and it's INT64MIN?

File: src/ConstantInterval.cpp [hunk 9]
```diff
@@ -0,0 +1,660 @@
...
+}
+
+ConstantInterval abs(const ConstantInterval &a) {
+    ConstantInterval result;
+    if (a.min_defined && a.max_defined && a.min != INT64_MIN) {
+        result.max_defined = true;
+        result.max = std::max(-a.min, a.max);
+    }
+    result.min_defined = true;
+    if (a.min_defined && a.min > 0) {
```
Comment (rootjalex) [inline src/ConstantInterval.cpp:577]: I think you also want a case for `(a.max_defined && a.max < 0)` -> `result.min = -a.max`

Comment (rootjalex) [reply src/ConstantInterval.cpp:577]: Probably also with a safety check that a.max != INT64_MIN

File: src/ConstantInterval.cpp [hunk 10]
```diff
@@ -0,0 +1,660 @@
...
+        result.min = a.min;
+    } else {
+        result.min = 0;
+    }
+
+    return result;
+}
+
+ConstantInterval operator<<(const ConstantInterval &a, const ConstantInterval &b) {
+    // Try to map this to a multiplication and a division
```
Comment (rootjalex) [inline src/ConstantInterval.cpp:587]: I'm not sure I follow this method entirely, more explanation in comments would be helpful

File: src/Monotonic.cpp [hunk 11]
```diff
@@ -493,10 +347,9 @@ class DerivativeBounds : public IRVisitor {
             return;
         }
 
-        if (op->is_intrinsic(Call::unsafe_promise_clamped) ||
-            op->is_intrinsic(Call::promise_clamped) ||
-            op->is_intrinsic(Call::saturating_cast)) {
+        if (op->is_intrinsic(Call::saturating_cast)) {
```
Comment (rootjalex) [inline src/Monotonic.cpp:350]: Why the removal of `(unsafe_)promise_clamped`?

**Review by rootjalex — APPROVED** [review]
Lgtm with nits

File: src/ConstantInterval.cpp (hunk 10 above)
Comment (abadams) [reply src/ConstantInterval.cpp:587]: It made sense to me at the time, but it doesn't now. I'll rewrite it.

File: src/Monotonic.cpp (hunk 11 above)
Comment (abadams) [reply src/Monotonic.cpp:350]: Because derivative_bounds calls remove_likelies and remove_promises before starting the visitor

**Comment by abadams** [comment]
I believe I have addresses review comments, PTAAL (mostly for operator<<)

File: src/ConstantInterval.cpp [hunk 12]
```diff
@@ -584,36 +588,51 @@ ConstantInterval abs(const ConstantInterval &a) {
...
+    // From there we can use the * and / operators.
+
+    ConstantInterval b_pos = max(b, 0), b_neg = max(-b, 0);
+
+    // At this point, we have sliced the interval b into two parts. E.g.
+    // if b = [10, 12],  b_pos = [10, 12] and b_neg = [0, 0]
+    // if b = [-4, 8],   b_pos = [0, 8]   and b_neg = [0, 4]
+    // if b = [-10, -3], b_pos = [0, 0]   and b_neg = [3, 10]
+    // if b = [-3, inf], b_pos = [0, inf] and b_neg = [0, 3]
+    // In all cases, note that b_pos - b_neg = b by our definition of - for
```
Comment (rootjalex) [inline src/ConstantInterval.cpp:612]: I think this is not true if b.min is defined and equal to INT64_MIN, because `(b_pos - b_neg).min` is not defined.

File: src/ConstantInterval.cpp [hunk 13]
```diff
@@ -584,36 +588,51 @@ ConstantInterval abs(const ConstantInterval &a) {
...
+
+    auto two_to_the = [](const ConstantInterval &i) {
+        const int64_t one = 1;
+        ConstantInterval r;
+        // We should know i is positive at this point.
+        internal_assert(i.min_defined && i.min >= 0);
+        r.min_defined = true;
+        if (i.min >= 63) {
+            // It's at least a value too large for us to represent, which is not
+            // the same as min_defined = false.
```
Comment (rootjalex) [inline src/ConstantInterval.cpp:623]: I'm not sure I understand what this distinction is exactly, and why it's used for `r.min` and not `r.max`

Comment (abadams) [reply src/ConstantInterval.cpp:623]: !min_defined means either there is no lower bound, or that the lower bound is too negative for us to represent. In either case, we know that, for example, -1000 is in the interval.

In this case there is a lower bound, but it's too *large* for us to represent. So -1000 is definitely not in the interval.

We don't have to worry about the same thing for max_defined, because this code is making large positive values.

Comment (rootjalex) [reply src/ConstantInterval.cpp:623]: Ah gotcha

File: src/ConstantInterval.cpp (hunk 12 above)
Comment (rootjalex) [reply src/ConstantInterval.cpp:612]: @abadams does this matter? If not, this PR lgtm

Comment (abadams) [reply src/ConstantInterval.cpp:612]: I had to think about it, so it warrants updating the comment. I've added: 

(ignoring corner cases, for which b_pos - b_neg safely over-approximates the bounds of b).

**Comment by abadams** [comment]
Well now there's a fuzz test failure, so something's probably not right with the new operator<<. Will investigate.

**Comment by abadams** [comment]
The fuzzer failure was just lossless_negate being wrong, which is to be fixed in #8155
